server/src/main/java/org/opensearch/index/store/Store.java

```diff
@@ -952,7 +954,11 @@ public void copyFrom(Directory from, String src, String dest, IOContext context)
             boolean success = false;
             long startTime = System.currentTimeMillis();
             try {
-                super.copyFrom(from, src, dest, context);
+                if (from instanceof RemoteSegmentStoreDirectory) {
+                    copyFileAndValidateChecksum(from, src, dest, context, fileSize);
+                } else {
+                    super.copyFrom(from, src, dest, context);
+                }
```

Comment:
Why do we need to add an if/else. Also fileSize can be derived from `from` and `src` so can we not implement the `copyFileAndValidateChecksum` to `RemoteSegmentStoreDirectory `

Reply:
> Why do we need to add an if/else?

This is a bit defensive to add checksum validation only for remote store downloads. I am not aware of all the flows where file is copied to StoreDirectory and whether it needs a stricter check. It is possible that caller is handling the checksum validation.

> Also fileSize can be derived from from and src so can we not implement the copyFileAndValidateChecksum to RemoteSegmentStoreDirectory 

Here, RemoteSegmentStoreDirectory is `from`. We can't use RemoteSegmentStoreDirectory's `copyFrom` method as in that method `from` would be some another directory.